### PR TITLE
Mjl index speedup

### PR DESCRIPTION
**Pull request by mjl**
This pull request addresses the problems when indexing large data sets, both in memory consumption and run time, as discussed in detail in issue #137.

Those changes have been running in a production environment for several months now without any problems.

### CODE REVIEW

**Comment by barseghyanartur** [comment]
@sabricot:

+1 for merge!

@mjl:

Thank you!

**Comment by pySilver** [comment]
any updates on this?

**Comment by barseghyanartur** [comment]
Come on, guys, this is one of the best improvements out there! It's a shame it's not merged yet.

**Comment by safwanrahman** [comment]
Sorry for taking so long. @mjl Can you remove the merge commit and rebase your changes upon master?

File: django_elasticsearch_dsl/documents.py [hunk 1]
```diff
@@ -44,6 +48,62 @@
...
+        small_cache_qs = self.qs.order_by('pk')
+
+        once = no_data = False
+        while not no_data and not once:
+            # If we got the max seen PK from last batch, use it to restrict the qs
+            # to values above; this optimises the query for Postgres as not to
+            # devolve into multi-second run time at large offsets.
+            if self.chunk_size:
+                if last_max_pk is not None:
+                    current_qs = small_cache_qs.filter(pk__gt=last_max_pk)[:self.chunk_size]
```
Comment (safwanrahman) [inline django_elasticsearch_dsl/documents.py]: What about when the `primary key` is not integer?

**Comment by safwanrahman** [comment]
I have gone through the code and it seems like it needs some improvement.

@barseghyanartur I understand your frustration, but this thing works for you does not mean will work for everyone. While maintaining packages, you need to make it generalized. So its not so easy to merge anything!

**Comment by barseghyanartur** [comment]
@safwanrahman:

I haven't seen breaking changes on the API level here. Everything stays on as it was.

**Comment by safwanrahman** [comment]
Its not breaking chnages in API level, but you need to ensure that it works properly for all the people who are using it. So its not a easy call

**Comment by mjl** [comment]
> What about when the `primary key` is not integer?

   It doesn't matter what type the primary key is, as long as it's 
sortable by the database. Integer, Strings, guuid, all should work fine.

   Also, that feature is opt-in, so if one really runs into problems, 
just don't enable it?

**Comment by safwanrahman** [comment]
@mjl You always can not ensure with `uuid` that the later one is always greater as its not incremental. Moreover, if its unique string, you always can not ensure that the later one will be greater. As you are making `gte` query, the later one will not be indexed.

**Comment by mjl** [comment]
> @mjl You always can not ensure with `uuid` that the later one is 
> always greater as its not incremental. Moreover, if its unique string, 
> you always can not ensure that the later one will be greater. As you 
> are making `gte` query, the later one will not be indexed.

   That is only a problem if you are adding data while doing a full 
reindex and expect the full reindex to include those entries. That, to 
be blunt, I would call user error, as iterating over changing data sets 
is always brittle. And note that won't even work as you expect with the 
current code: Anything that is added while the index is rebuilt will not 
be included in the index as the sql query is done at a certain point in 
time (and thus its result set frozen); anything added after that time 
will not be indexed by the rebuild command. You might catch that case by 
using signals or whatnot, but that is outside the scope of the indexing 
code.

   But fair enough, perhaps let's add some warning in the docs? What 
should it say?

**Comment by safwanrahman** [comment]
> That is only a problem if you are adding data while doing a full
reindex and expect the full reindex to include those entries. That, to
be blunt, I would call user error, as iterating over changing data sets
is always brittle. And note that won't even work as you expect with the
current code: Anything that is added while the index is rebuilt will not
be included in the index as the sql query is done at a certain point in
time (and thus its result set frozen); anything added after that time
will not be indexed by the rebuild command. You might catch that case by
using signals or whatnot, but that is outside the scope of the indexing
code.

   But fair enough, perhaps let's add some warning in the docs? What
should it say?

@mjl I understand your point. I think adding documentation will be better idea.

I have a feeling to not change the `get_queryset` method. As its a public method, we should not change it.
According to [documentation](https://docs.djangoproject.com/en/2.2/ref/models/querysets/), you could use `iterator` with `chunk_size`. Django would manage everything that you are going to do actually. Either it will use the database level cursor, or load it in memory.

File: django_elasticsearch_dsl/documents.py [hunk 2]
```diff
@@ -44,6 +48,62 @@
...
+                else:
+                    current_qs = small_cache_qs[:self.chunk_size]
```
Comment (safwanrahman) [inline django_elasticsearch_dsl/documents.py]: Whats the reason to slice before running the `iterator`? [Django handle it by default if you pass certain ](https://docs.djangoproject.com/en/2.2/ref/models/querysets/) chunk_size`

File: django_elasticsearch_dsl/documents.py [hunk 3]
```diff
@@ -134,16 +217,16 @@ def _prepare_action(self, object_instance, action):
...
-            for page in paginator.page_range:
-                for object_instance in paginator.page(page).object_list:
-                    yield self._prepare_action(object_instance, action)
+        for object_instance in object_list:
+            yield self._prepare_action(object_instance, action)
+
+    def _bulk(self, *args, **kwargs):
+        """Helper for switching between normal and parallel bulk operation"""
+        parallel = kwargs.pop('parallel', False)
+        if parallel:
```
Comment (safwanrahman) [inline django_elasticsearch_dsl/documents.py]: We should find a way where while updating a single object the `parallel` is always false. We need to use it only while indexing large number of objects. When a model object is updated, the django signal call the same method for updating the index.

**Comment by mjl** [comment]
> +            # If we got the max seen PK from last batch, use it to 
> restrict the qs
> +            # to values above; this optimises the query for Postgres 
> as not to
> +            # devolve into multi-second run time at large offsets.
> +            if self.chunk_size:
> +                if last_max_pk is not None:
> +                    current_qs = 
> small_cache_qs.filter(pk__gt=last_max_pk)[:self.chunk_size]
> +                else:
> +                    current_qs = small_cache_qs[:self.chunk_size]


> Whats the reason to slice before running the `iterator`? [Django 
> handle it by default if you pass certain 
> ](https://docs.djangoproject.com/en/2.2/ref/models/querysets/) 
> chunk_size`

   Somewhere in the interaction between django and postgres, iterating 
over large datasets produces very (VERY) long query times. In my tests, 
the first couple of batches came in at 200ms per query, and later 
batches came in at 12000ms. That problem might or might not have been 
fixed in the meantime, I really do not know. The code has originated in 
django-haystack in 2017, so things might be better now (one can always 
hope!).

   I have analyzed the problem in 
https://github.com/django-haystack/django-haystack/issues/1077#issuecomment-277677653

**Comment by mjl** [comment]
> We should find a way where while updating a single object the 
> `parallel` is always false. We need to use it only while indexing 
> large number of objects. When a model object is updated, the django 
> signal call the same method for updating the index.

   Isn't the _bulk thingy only used by the reindex command?

**Comment by safwanrahman** [comment]
> Isn't the _bulk thingy only used by the reindex command?

unfortunately no! I also wish the `_bulk` would be called in reindex only, but it seems like while indexing a single object, `_bulk` is called. 😢

**Comment by safwanrahman** [comment]
> Somewhere in the interaction between django and postgres, iterating
over large datasets produces very (VERY) long query times. In my tests,
the first couple of batches came in at 200ms per query, and later
batches came in at 12000ms. That problem might or might not have been
fixed in the meantime, I really do not know. The code has originated in
django-haystack in 2017, so things might be better now (one can always
hope!).

I think the problem might have fixed in the meantime. I have use `iterator` in my workplace for querying a table hacing millions of data, but did not face any problem. Let me know if the issue still persist so I can investigate more.

**Comment by mjl** [comment]
> I think the problem might have fixed in the meantime. I have use 
> `iterator` in my workplace for querying a table hacing millions of 
> data, but did not face any problem.

   Ok, you mean just remove the slicing action and use the iterator 
directly? I can try that and compare it, no problem.

**Comment by mjl** [comment]
> unfortunately no! I also wish the `_bulk` would be called in reindex 
> only, but it seems like while indexing a single object, `_bulk` is 
> called. 😢

   Ok, let's see if we can distinguish those cases...

**Comment by safwanrahman** [comment]
> Ok, you mean just remove the slicing action and use the iterator
directly? I can try that and compare it, no problem.

Yes. I think we can remove the `PagingQuerysetProxy` object and handle it directly in the `Document`. what do you think?

**Comment by mjl** [comment]
> Yes. I think we can remove the `PagingQuerysetProxy` object and handle 
> it directly in the `Document`. what do you think?

   I'll try and report back. I'm all for removing unnecessary 
complexity!

**Comment by pySilver** [comment]
@mjl sry to bother you :) any news from your side?

**Comment by mjl** [comment]
> @mjl sry to bother you :) any news from your side?

   Yes, thanks for the poke. I've done some measurements that I need to 
bring into understandable format, but in a nutshell, it really looks as 
if the paging part is no longer necessary. I'll summarize what I have in 
a bit.

**Comment by pySilver** [comment]
@mjl actually there is some issue when using non standard queryset:

```python
DISTINCT_BY_FIELDS = [
        "item_group_id",
        "brand_id",
        "merchant_id",
        "material",
        "pattern",
        "color",
        "adult",
        "multipack",
        "is_bundle",
        "age_group",
        "gender",
        "size_type",
        "size_system",
    ]
    def get_queryset(self):
        # In case of performance issues we should use `CROSS  JOIN LATERAL`
        # @see: https://medium.com/kkempin/postgresqls-lateral-join-bfd6bd0199df
        # @see: https://heap.io/blog/engineering/postgresqls-powerful-new-join-type-lateral
        # @see: https://dbfiddle.uk/?rdbms=postgres_11&fiddle=68cc26893cb96a912997ac6342f42490
        return (
            super()
            .get_queryset()
            .select_related("brand", "merchant", "coupon")
            .distinct(*self.DISTINCT_BY_FIELDS)
            .order_by(*self.DISTINCT_BY_FIELDS)
        )
```

```Are you sure you want to delete the '<elasticsearch_dsl.index.Index object at 0x10c481190>' indexes? [n/Y]: y
Deleting index '<elasticsearch_dsl.index.Index object at 0x10c481190>'
INFO 2019-10-01 16:33:42,030 base 90395 4591814080 DELETE http://localhost:9200/products [status:200 request:0.039s]
Creating index '<elasticsearch_dsl.index.Index object at 0x10c481190>'
INFO 2019-10-01 16:33:42,108 base 90395 4591814080 PUT http://localhost:9200/products [status:200 request:0.075s]
Indexing 1003 'Product' objects
Traceback (most recent call last):
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.InvalidColumnReference: SELECT DISTINCT ON expressions must match initial ORDER BY expressions
LINE 1: ...O SCROLL CURSOR WITH HOLD FOR SELECT DISTINCT ON ("catalog_p...
                                                             ^


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django/db/models/sql/compiler.py", line 1100, in execute_sql
    cursor.execute(sql, params)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django/db/backends/utils.py", line 99, in execute
    return super().execute(sql, params)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django/db/backends/utils.py", line 67, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django/db/backends/utils.py", line 76, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django/db/utils.py", line 89, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: SELECT DISTINCT ON expressions must match initial ORDER BY expressions
LINE 1: ...O SCROLL CURSOR WITH HOLD FOR SELECT DISTINCT ON ("catalog_p...
                                                             ^


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./manage.py", line 30, in <module>
    execute_from_command_line(sys.argv)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django/core/management/__init__.py", line 375, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django/core/management/base.py", line 323, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django/core/management/base.py", line 364, in execute
    output = self.handle(*args, **options)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django_elasticsearch_dsl/management/commands/search_index.py", line 134, in handle
    self._rebuild(models, options)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django_elasticsearch_dsl/management/commands/search_index.py", line 115, in _rebuild
    self._populate(models, options)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django_elasticsearch_dsl/management/commands/search_index.py", line 92, in _populate
    doc().update(qs)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django_elasticsearch_dsl/documents.py", line 248, in update
    **kwargs
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django_elasticsearch_dsl/documents.py", line 227, in _bulk
    return self.parallel_bulk(*args, **kwargs)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django_elasticsearch_dsl/documents.py", line 206, in parallel_bulk
    deque(parallel_bulk(client=self._get_connection(), actions=actions, **kwargs), maxlen=0)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/elasticsearch/helpers/actions.py", line 365, in parallel_bulk
    actions, chunk_size, max_chunk_bytes, client.transport.serializer
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/multiprocessing/pool.py", line 748, in next
    raise value
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/multiprocessing/pool.py", line 121, in worker
    result = (True, func(*args, **kwds))
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/multiprocessing/pool.py", line 140, in _helper_reraises_exception
    raise ex
  File "/usr/local/Cellar/python/3.7.4/Frameworks/Python.framework/Versions/3.7/lib/python3.7/multiprocessing/pool.py", line 292, in _guarded_task_generation
    for i, x in enumerate(iterable):
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/elasticsearch/helpers/actions.py", line 63, in _chunk_actions
    for action, data in actions:
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django_elasticsearch_dsl/documents.py", line 220, in _get_actions
    for object_instance in object_list:
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django_elasticsearch_dsl/documents.py", line 100, in __iter__
    for obj in current_qs.iterator():
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django/db/models/query.py", line 341, in _iterator
    yield from self._iterable_class(self, chunked_fetch=use_chunked_fetch, chunk_size=chunk_size)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django/db/models/query.py", line 55, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch, chunk_size=self.chunk_size)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django/db/models/sql/compiler.py", line 1103, in execute_sql
    cursor.close()
psycopg2.errors.InvalidCursorName: cursor "_django_curs_123145595777024_1" does not exist```

**Comment by mjl** [comment]
Ah I see. As the PagingProxy tacks its own order_by() onto the 
queryset, that of course does no longer match the distinct() you add in 
get_queryset(), and thus it blows up with

> psycopg2.errors.InvalidColumnReference: SELECT DISTINCT ON expressions 
> must match initial ORDER BY expressions

   On the positive side, all that manual slicing seems unnecessary in 
recent djangos/psycopg2. Thanks for the report -- may I ping you for a 
new version to try later?

   In the meantime, you can try to do the following: In 
documents.py:135, make get_indexing_queryset() just return 
self.get_queryset().iterator(). That's untested, but should disable the 
PagingQuerysetProxy altogether and be pretty much what I'm thinking 
should work just as well.

**Comment by pySilver** [comment]
@mjl sure, thanks for heads up! if you really need specific ordering you can append a column to existing ordering fields (postgresql requires prefix of distinct and orderby to be the same)

**Comment by pySilver** [comment]
Returning queryset.iterator() is not working due to:

```
INFO 2019-10-01 17:08:15,979 base 90722 4596716992 PUT http://localhost:9200/products [status:200 request:0.084s]
Traceback (most recent call last):
  File "./manage.py", line 30, in <module>
    execute_from_command_line(sys.argv)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django/core/management/__init__.py", line 381, in execute_from_command_line
    utility.execute()
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django/core/management/__init__.py", line 375, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django/core/management/base.py", line 323, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django/core/management/base.py", line 364, in execute
    output = self.handle(*args, **options)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django_elasticsearch_dsl/management/commands/search_index.py", line 134, in handle
    self._rebuild(models, options)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django_elasticsearch_dsl/management/commands/search_index.py", line 115, in _rebuild
    self._populate(models, options)
  File "/Users/Silver/.local/share/virtualenvs/mybaze-dsw6HN8p/lib/python3.7/site-packages/django_elasticsearch_dsl/management/commands/search_index.py", line 90, in _populate
    qs.count(), doc.django.model.__name__)
AttributeError: 'generator' object has no attribute 'count'
```

**Comment by pySilver** [comment]
Plain `return self.get_queryset()` works.

**Comment by pySilver** [comment]
In order to make it work with `PagingQuerysetProxy`  the following should be changed:

at `django_elasticsearch_dsl.documents.PagingQuerysetProxy.__iter__`

from:
`small_cache_qs = self.qs.order_by('pk')`
to:
`small_cache_qs = self.qs.order_by(self.qs.query.order_by + ('pk',))`
(so it preserves ordering provided in orignal qs if any)

**Comment by mjl** [comment]
No, using a different ordering will not work correctly. And not using iterator will consume unbelievable amounts of memory, so you miss out of most of the benefits.

**Comment by pySilver** [comment]
hm why different ordering won't work?

**Comment by mjl** [comment]
Because the PagingQuerysetProxy chops up the result set ordered by pk. If your queryset is not ordered by pk, then a full reindex will certainly not index the whole dataset. That's the reason the order_by('pk') that caused your exception in the first place is there.

I'll keep you updated when I have something new.

**Comment by pySilver** [comment]
im not 100% sure but as I understand `distinct on` in my case it should not have any difference. results are sorted by pk, preceding ordering keys are needed for db to eliminate duplicates. but again, I cannot test it.

**Comment by mjl** [comment]
Okay, I have a small analysis of the state of affairs here...

### Plain ded 6.4.2

    Deleting index 'kit-20191002100942'
    Creating index 'kit-20191002100942'
    Indexing 406304 'Plastic' objects

    real	31m12.357s
    user	5m59.528s
    sys	0m10.205s

### Plain ded 6.4.2 plus .iterator() in _populate

```
Deleting index 'kit-20191002104856'
Creating index 'kit-20191002104856'
Indexing 406304 'Plastic' objects

real	12m25.691s
user	5m56.511s
sys	0m8.887s
```

### ded mjl-index-speedup-6.4.2

```
Deleting index 'kit-20191002110406'
Creating index 'kit-20191002110406'
Indexing 406304 'Plastic' objects

real	6m56.811s
user	4m27.797s
sys	0m7.669s
```

### ded mjl-index-speedup-6.4.2 with PagingQuerysetProxy ripped out

```
Deleting index 'kit-20191002114453'
Creating index 'kit-20191002114453'
Indexing 406304 'Plastic' objects

real	6m4.546s
user	4m27.598s
sys	0m9.618s
```

### Same, but remove order_by from model.Meta

```
Deleting index 'kit-20191002122409'
Creating index 'kit-20191002122409'
Indexing 406304 'Plastic' objects

real	5m26.150s
user	4m26.708s
sys	0m8.720s
```

### Memory profiling

Figure_1

Also interesting is the memory consumption: as you can see, iterator() reduces memory pressure considerably. And also, you see a flat "step" in most of the graphs, that is the database sorting the result set, thus my last test without an model.Meta.ordering.

@pySilver, can you try the latest from the mjl-index-speedup-6 branch? That should be at ccf4666 right now? I have not updated the -7 branch yet, because I don't have an ES7 to test with.

**Comment by pySilver** [comment]
@mjl wow, thats a nice analysis you've made!

Unfortunately I cannot test on 6x branch since we are using 7x :/ I'll take a look at changes and apply to 7.x

**Comment by mjl** [comment]
> Unfortunately I cannot test on 6x branch since we are using 7x :/ I'll 
> take a look at changes and apply to 7.x

   Ok, that was easy, try mjl-index-speedup-7. It's a bit behind the 
official ded-7, but might be enough to get you going?

**Comment by pySilver** [comment]
Yes, I'll handle this in a few mins. Is there anything you'd like me to check particularly?

**Comment by pySilver** [comment]
if there a bench script – I can run it

**Comment by mjl** [comment]
No, just try whatever you feel you need to try. You seem to have pretty involved model structure and queries, so if it works for you, it probably works for pretty much everybody else :-)

Let me know how it goes, and if it helps, speedwise. Thanks!

**Comment by pySilver** [comment]
@mjl it work perfectly, thank you!

@sabricot isn't it ready to be merged into master?

**Comment by safwanrahman** [comment]
@pySilver I have given some fixup for this PR, once its fixed, it will be able to merge! :D

**Comment by safwanrahman** [comment]
@mjl do you need any kind of support to fix the issues I have mentioned? Feel free to let me know.

**Comment by pySilver** [comment]
@mjl @safwanrahman  

I've checked current code and 2 out of 3 issues mentioned by @safwanrahman looks outdated at the moment. The one left is about `django_elasticsearch_dsl.documents.DocType._bulk` which apparently is called for single item's updates (signals). Is it still relevant? Maybe simply propagate some flag from management command there?

**Comment by mjl** [comment]
> The one left is about 
> `django_elasticsearch_dsl.documents.DocType._bulk`

That's a coincidence, I just looked into that... _bulk() only switches 
to parallel mode if a parallel=True kwarg is present. This comes through 
update() and is copied there from the document's options. In order to 
make that only set for a full reindex, I'd have to pass that parameter 
down from management/commands/search_index _populate(), which sounds not 
too bad.

Basically, three smallish changes. Would that do the trick @sabricot?

```
@@ -172,7 +172,7 @@ class DocType(DSLDocument):
          else:
              return self.bulk(*args, **kwargs)

-    def update(self, thing, refresh=None, action='index', **kwargs):
+    def update(self, thing, refresh=None, action='index', 
may_parallelise=False, **kwargs):
          """
          Update each document in ES for a model, iterable of models or 
queryset
          """
@@ -188,7 +188,7 @@ class DocType(DSLDocument):

          return self._bulk(
              self._get_actions(object_list, action),
-            parallel=self.django.parallel_indexing,
+            parallel=may_parallelise and self.django.parallel_indexing,
              **kwargs
          )

diff --git 
a/django_elasticsearch_dsl/management/commands/search_index.py 
b/django_elasticsearch_dsl/management/commands/search_index.py
index e4cc7ef..00d25a9 100644
--- a/django_elasticsearch_dsl/management/commands/search_index.py
+++ b/django_elasticsearch_dsl/management/commands/search_index.py
@@ -89,7 +89,7 @@ class Command(BaseCommand):
                  doc().get_queryset().count(), 
doc.django.model.__name__)
              )
              qs = doc().get_indexing_queryset()
-            doc().update(qs)
+            doc().update(qs, may_parallelise=True)

```

**Comment by safwanrahman** [comment]
@mjl I think it can be done. But I have mentioned couple of issues above that need to be fixed.
@pySilver There are some more issue to fix like removing the slicing and remove the extreme complexity.

**Comment by safwanrahman** [comment]
I think its better to pass a argument while reindexing rathar than setting parallel index in document level.

**Comment by mjl** [comment]
> @mjl I think it can be done. But I have mentioned couple of issues 
> above that need to be fixed.

   I went over this thread and all I think all the issues you mentioned 
have been resolved (by ripping out the PagingQuerysetProxy). Did I miss 
some?

**Comment by mjl** [comment]
> I think its better to pass a argument while reindexing rathar than 
> setting parallel index in document level.

But then you would have to set that argument every time you do a 
reindex? I think a parameter makes sense in addition (so you can "try it 
out" without having to change things), but it's kinda nice you can set a 
default in the document...?

**Comment by safwanrahman** [comment]
@mjl Its only used the time when anyone run `populate` or `reindex` management command. I think, it does not make sense to have it in the document level because its not related to document. If we add the configuration in document level, people may think that its used while indexing single object also, but according to our approach, its not the case.

**Comment by safwanrahman** [comment]
@mjl Following work is needed actually, then I can give another review.
- Removing `PagingQuerysetProxy`
- Fixing the queryset slicing
- management command argument for paraller indexing

**Comment by mjl** [comment]
Okay, I see you point. Still I like the "set a default and forget" 
convenience; so perhaps rename it to parallel_only_used_on_full_reindex 
:-) or somesuch?

**Comment by mjl** [comment]
> - Removing `PagingQuerysetProxy`
> - Fixing the queryset slicing

Those are done.

> - management command argument for paraller indexing

Give me a minute :-)

**Comment by safwanrahman** [comment]
> Okay, I see you point. Still I like the "set a default and forget"
convenience; so perhaps rename it to parallel_only_used_on_full_reindex
:-) or somesuch?

What about adding a value in settings where people can add the default one? I think adding a settings makes more sense if you want something like add one time and forget! 😉

**Comment by mjl** [comment]
Fair enough! I've added ELASTICSEARCH_DSL_PARALLEL to settings and 
--parallel to the search_index management command.

Let me know if anything else needs adjusting!

**Comment by safwanrahman** [comment]
@mjl Have you pushed your changes? I dont see it updated!

**Comment by safwanrahman** [comment]
I think you have pushed to other branch, not this branch!

**Comment by mjl** [comment]
Ah sorry, I'll incorporate the changes in this branch in a sec.

**Comment by pySilver** [comment]
I agree with @sabricot that parallel does not belong on document level but management command. And I think it should be optional  flag to follow what ES does (it does not imply bulk indexing by default). Having a config flag for "set and forget" sounds as a good idea.

**Comment by safwanrahman** [comment]
@pySilver @safwanrahman and @sabricot are different person! 😅

**Comment by mjl** [comment]
I had to delete and re-create my branch because it would not rebase easily;
a new pull request was necessary -> #213

**Pull request closed without merging**